llmfoundry/models/hf/hf_base.py

```diff
@@ -331,53 +334,111 @@ def build_inner_model(
+            if resolved_init_device == 'cpu':
+                if pretrained:
+                    model = auto_model_cls.from_pretrained(
+                        pretrained_model_name_or_path,
+                        trust_remote_code=trust_remote_code,
+                        use_auth_token=use_auth_token,
+                        load_in_8bit=load_in_8bit,
+                        attn_implementation=requested_attention_implementation,
+                        config=config,
+                    )
+                else:
+                    model = auto_model_cls.from_config(  # type: ignore
+                        config,
+                        trust_remote_code=trust_remote_code,
+                        attn_implementation=requested_attention_implementation,
+                    )
+            elif resolved_init_device == 'meta':
+                if pretrained:
+                    raise ValueError(
+                        'Setting cfg.pretrained=True is not supported when init_device="meta".',
+                    )
+                with init_empty_weights(include_buffers=False):
+                    model = auto_model_cls.from_config(  # type: ignore
+                        config,
+                        trust_remote_code=trust_remote_code,
+                        attn_implementation=requested_attention_implementation,
+                    )
             else:
-                model = auto_model_cls.from_config(  # type: ignore
-                    config,
-                    trust_remote_code=trust_remote_code,
-                    attn_implementation=requested_attention_implementation,
-                )
-        elif resolved_init_device == 'meta':
-            if pretrained:
                 raise ValueError(
-                    'Setting cfg.pretrained=True is not supported when init_device="meta".',
-                )
-            with init_empty_weights(include_buffers=False):
-                model = auto_model_cls.from_config(  # type: ignore
-                    config,
-                    trust_remote_code=trust_remote_code,
-                    attn_implementation=requested_attention_implementation,
+                    f'init_device="{init_device}" must be either "cpu" or "meta".',
                 )
-        else:
-            raise ValueError(
-                f'init_device="{init_device}" must be either "cpu" or "meta".',
+        except Exception as e:
+            error = e
+            log.error(e)
+
+        # Signal that the download is complete on rank 0
+        if download_status_queue is not None:
+            download_status_queue.put(1)
+            log.debug('Download complete')
+
+        # Join and wait for the thread to complete on all ranks
+        if download_thread is not None:
+            log.debug('Joining download thread')
+            download_thread.join(timeout=3600)
+            log.debug('Download thread joined')
+
+        # Gather information about which ranks errored
```

Comment:
So if init_device=cpu or meta (also true for mixed but with mixed we have the assumption that rank 0 takes a long time compared to other ranks), we would first wait for rank 0 to finish and then be subject to the dist timeout here.

Reply:
seems fine, just thinking aloud

Reply:
yup. seems like there could be an edge case there with cpu init, but the ranks took wildly different amounts of time to download and rank 0 finished more than a `dist_timeout` ahead, and then you would still timeout. I think thats ok though, can handle later if it comes up for some reason, but should really always be using mixed init here.